neard/src/runtime.rs

```diff
@@ -1042,11 +1042,18 @@ impl RuntimeAdapter for NightshadeRuntime {
 
     fn obtain_state_part(&self, state_root: &StateRoot, part_id: u64, num_parts: u64) -> Vec<u8> {
         assert!(part_id < num_parts);
-        self.trie
-            .get_trie_nodes_for_part(part_id, num_parts, state_root)
-            .expect("storage should not fail")
-            .try_to_vec()
-            .expect("serializer should not fail")
+        match self.trie.get_trie_nodes_for_part(part_id, num_parts, state_root) {
+            Ok(partial_state) => partial_state,
+            Err(e) => {
+                error!(target: "runtime",
+                    "Can't get_trie_nodes_for_part for {:?}, part_id {:?}, num_parts {:?}, {:?}",
+                    state_root, part_id, num_parts, e
+                );
+                panic!("RuntimeError::StorageInconsistentState, {:?}", e)
```

Comment:
why not have everything in `panic`?

Reply:
I see no meaningful difference.
BTW, I'm not sure that a node should stop working if it can't create a State Part.

Reply:
Good point. So is `panic` added here just for debugging? If it is not needed we can remove it

Reply:
No, it panics previously. I just copied that behavior.

Reply:
I think we should remove it unless you have objections.

Reply:
before the change there is an `unwrap()`, panic preserves existing behavior.

Reply:
I understand that part, but why panicking here at all?

Reply:
Let's leave panic for now because we want to emphasize the issue.

Reply:
Because we detected that storage is corrupted. We shouldn't continue execution if there is a bug or a database failure.

Reply:
`obtain_state_part` is called by `get_state_response_part`, which explicitly checks that state root node is there:
```        
let state_root_node = self.runtime_adapter.get_state_root_node(&state_root);
```

`get_trie_nodes_for_part` failure means that some of the nodes aren't there.
This is a capital violation of trie invariant.

If instead of state sync someone asks us to validate a block on top of this state or query it, we will run into the same issue.
Panic here allows us to find the bug sooner.

Reply:
@mikhailOK how does it possible if `state_root_node` exists but some leaves are not? can it appear because we added some key-value (K,V) on height A, then added same (K,V) on height B and then GC height A and removed (K,V) but B still needs it?

Reply:
either that or it actually doesn't exist but there is also a cache inconsistency

Reply:
I see

Reply:
Actually in that case let's either move this function into where it is called or document this function with the note that it should only be called assuming we have state_root_node so that we don't accidentally call it in some other place that might crash the node.